src/tools/perf/lib/ucp_tests.cc

```diff
@@ -1033,35 +1033,69 @@ class ucp_perf_test_runner {
 {
     UCS_PP_FOREACH(TEST_CASE_ALL_OSD, perf,
-        (UCX_PERF_CMD_PUT,   UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_PUT,   UCX_PERF_TEST_TYPE_PINGPONG_WAIT_MEM),
-        (UCX_PERF_CMD_PUT,   UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_GET,   UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_ADD,   UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_FADD,  UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_SWAP,  UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_CSWAP, UCX_PERF_TEST_TYPE_STREAM_UNI)
-        );
+                   (UCX_PERF_CMD_PUT, UCX_PERF_TEST_TYPE_PINGPONG),
+                   (UCX_PERF_CMD_PUT, UCX_PERF_TEST_TYPE_PINGPONG_WAIT_MEM),
+                   (UCX_PERF_CMD_PUT, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_GET, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_ADD, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_FADD, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_SWAP, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_CSWAP, UCX_PERF_TEST_TYPE_STREAM_UNI)
+                   );
+    return UCS_ERR_INVALID_PARAM;
+}
 
+static ucs_status_t dispatch_tag(ucx_perf_context_t *perf)
+{
     UCS_PP_FOREACH(TEST_CASE_ALL_TAG, perf,
-        (UCX_PERF_CMD_TAG,      UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_TAG,      UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_TAG_SYNC, UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_TAG_SYNC, UCX_PERF_TEST_TYPE_STREAM_UNI)
-        );
+                   (UCX_PERF_CMD_TAG, UCX_PERF_TEST_TYPE_PINGPONG),
+                   (UCX_PERF_CMD_TAG, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_TAG_SYNC, UCX_PERF_TEST_TYPE_PINGPONG),
+                   (UCX_PERF_CMD_TAG_SYNC, UCX_PERF_TEST_TYPE_STREAM_UNI)
+                   );
+    return UCS_ERR_INVALID_PARAM;
+}
 
+static ucs_status_t dispatch_stream(ucx_perf_context_t *perf)
+{
     UCS_PP_FOREACH(TEST_CASE_ALL_STREAM, perf,
-        (UCX_PERF_CMD_STREAM,   UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_STREAM,   UCX_PERF_TEST_TYPE_PINGPONG)
-        );
+                   (UCX_PERF_CMD_STREAM, UCX_PERF_TEST_TYPE_STREAM_UNI),
+                   (UCX_PERF_CMD_STREAM, UCX_PERF_TEST_TYPE_PINGPONG)
+                   );
+    return UCS_ERR_INVALID_PARAM;
+}
 
+static ucs_status_t dispatch_am(ucx_perf_context_t *perf)
+{
     UCS_PP_FOREACH(TEST_CASE_ALL_AM, perf,
-        (UCX_PERF_CMD_AM,       UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_AM,       UCX_PERF_TEST_TYPE_STREAM_UNI)
-        );
+                   (UCX_PERF_CMD_AM, UCX_PERF_TEST_TYPE_PINGPONG),
+                   (UCX_PERF_CMD_AM, UCX_PERF_TEST_TYPE_STREAM_UNI)
+                   );
+    return UCS_ERR_INVALID_PARAM;
+}
+
+typedef ucs_status_t (*ucp_dispatch_func_t)(ucx_perf_context_t *perf);
+
+static ucp_dispatch_func_t dispatchers[] = {
```

Comment:
IMO better use std::vector here

Reply:
done

Reply:
Actually, it won't compile with a vector.
The reason we can't use a vector is that during the linking phase (CCLD ucx_perftest), gcc is used, which leads to missing references for operator new and operator delete:
usr/bin/ld.bfd: lib/.libs/libucxperf.a(libucxperf_la-ucp_tests.o): in function `__gnu_cxx::new_allocator<ucs_status_t (*)(ucx_perf_context*)>::allocate(unsigned long, void const*)': <--[gcc]`

Reply:
please check again @Artemy-Mellanox